plugins/reboot/reboot.go

```diff
@@ -50,24 +102,51 @@ func (plugin) NewTaskPlugin(options plugins.TaskPluginOptions) (plugins.TaskPlug
 		return nil, engines.ErrContractViolation
 	}
 
-	return taskPlugin{
+	pl.runningTasks.RLock()
+	return &taskPlugin{
 		TaskPluginBase: plugins.TaskPluginBase{},
-		reboot:         p.Reboot,
+		plugin:         pl,
+		rebootWhenDone: p.RebootWhenDone,
 	}, nil
 }
 
-func (tp taskPlugin) Dispose() error {
-	if tp.reboot {
-		if err := reboot(); err != nil {
-			return engines.NewInternalError(err.Error())
-		}
+func (tp *taskPlugin) Dispose() error {
+	tp.plugin.runningTasks.RUnlock()
+
+	// Note: in a multi-task environment, this will not reboot immediately, but
+	// will wait until all tasks are complete.
+	if tp.rebootWhenDone {
+		tp.plugin.environment.Log.Warning("Task specified reboot after completion; scheduling reboot")
+		tp.plugin.reboot()
 	}
 
 	return nil
 }
 
 func (pluginProvider) NewPlugin(options plugins.PluginOptions) (plugins.Plugin, error) {
-	return plugin{}, nil
+	var c config
+	if err := schematypes.MustMap(configSchema, options.Config, &c); err != nil {
+		return nil, engines.ErrContractViolation
+	}
+
+	plugin := &plugin{
+		PluginBase:   plugins.PluginBase{},
+		environment:  options.Environment,
+		runningTasks: sync.RWMutex{},
+	}
+
+	if c.RebootAfter > 0 {
+		// after rebootAfter hours, call the reboot method
+		go func() {
+			select {
+			case <-time.After(time.Duration(c.RebootAfter) * time.Hour):
+				plugin.environment.Log.Warning("Host rebootAfter has expired; scheduling reboot")
+				plugin.reboot()
+			}
+		}()
+	}
```

Comment:
hrm, shouldn't this go inside `pluginProvider`? I think `NewPlugin` can be called multiple times during worker life time, and only when a task is created.

@jonasfj you know better the ins and outs of task starting flow, am I right here?